File: AdvancePython/Day-4 Modular Programming/Task-5.py
# Task 5:
import time
import functools


def timer(func: callable) -> callable:

    @functools.wraps(func)
    def wrapper(*args: any, **kwargs: any) -> any:
        start_time = time.perf_counter()
        result = func(*args, **kwargs)
        end_time = time.perf_counter()
        print(f"Time taken: {end_time - start_time:.4f} seconds.")
        return result

    return wrapper


@timer
def waste_time() -> None:
    time.sleep(1)
    print("Function finished!")


waste_time()

File: node.js/Day-4/Task-5.js
// Task 5:
function timer(func) {
    return async function(...args) {
        const startTime = process.hrtime.bigint(); // start time
        const result = await func(...args); // Run the original function
        const endTime = process.hrtime.bigint(); // end time

        const seconds = Number(endTime - startTime) / 1_000_000_000;
        console.log(`Time taken: ${seconds.toFixed(4)} seconds.`);
    };
}

const sleep = (ms) => new Promise(resolve => setTimeout(resolve, ms));

async function wasteTime() {
    await sleep(1000); // Wait for 1 second
    console.log("Function finished!");
}

const timedWasteTime = timer(wasteTime);

timedWasteTime();